content/en/blog/_posts/2025-04-11-sidecar-patterns/index.md

```diff
@@ -0,0 +1,89 @@
+---
+layout: blog
+title: "Kubernetes Sidecar Patterns: An Overview"
+date: 2025-04-11
+draft: false 
+slug: sidecar-patterns
+author: Agata Skorupka (The Scale Factory)
+---
+
+As cloud-native architectures continue to evolve, Kubernetes has become the go-to platform for deploying complex, distributed systems. One of the most powerful yet nuanced design patterns in this ecosystem is the sidecar pattern—a technique that allows developers to extend application functionality without diving deep into source code.
+
+## The origins of the sidecar pattern
+
+Think of a sidecar like a trusty companion motorcycle attachment. Historically, IT infrastructures have always used auxiliary services to handle critical tasks. Before containers, we relied on background processes and helper daemons to manage logging, monitoring, and networking. The microservices revolution transformed this approach, making sidecars a structured and intentional architectural choice.
+With the rise of microservices, the sidecar pattern became more clearly defined, allowing developers to offload specific responsibilities from the main service without altering its code. Service meshes like Istio and Linkerd have popularized sidecar proxies, demonstrating how these companion containers can elegantly handle observability, security, and traffic management in distributed systems.
+
+## Kubernetes implementation
+
+In Kubernetes, [sidecar containers](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) operate within the same pod as the main application, enabling communication and resource sharing. Does this sound just like defining multiple containers along each other inside the pod? It actually does, and this is how sidecar containers had to be implemented before Kubernetes v1.29.0, which introduced native support for sidecars, which can now be defined within a Kubernetes resources using the `spec.initContainers` field, but with `restartPolicy: Always` as in the example below, which is a partial snippet of the full Kubernetes manifest:
```

Comment:
In Kubernetes, [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/) operate within
the same pod as the main application, enabling communication and resource sharing.
Does this sound just like defining multiple containers along each other inside the pod? It actually does, and
this is how sidecar containers had to be implemented before Kubernetes v1.29.0, which introduced
native support for sidecars.
Sidecar containers  can now be defined within a Pod manifest using the `spec.initContainers` field. What makes
it a sidecar container is that you specify it with `restartPolicy: Always`. You can see an example of this below, which is a partial snippet of the full Kubernetes manifest:
```